tracer/src/Datadog.Trace/Configuration/ImmutableDynamicSettings.cs

```diff
@@ -0,0 +1,109 @@
+// <copyright file="ImmutableDynamicSettings.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+
+namespace Datadog.Trace.Configuration
+{
+    internal class ImmutableDynamicSettings : IEquatable<ImmutableDynamicSettings>
+    {
+        public bool? RuntimeMetricsEnabled { get; init; }
+
+        public bool? DataStreamsMonitoringEnabled { get; init; }
+
+        public string? CustomSamplingRules { get; init; }
+
+        public double? GlobalSamplingRate { get; init; }
+
+        public string? SpanSamplingRules { get; init; }
+
+        public bool? LogsInjectionEnabled { get; init; }
+
+        public IReadOnlyDictionary<string, string>? HeaderTags { get; init; }
+
+        public IReadOnlyDictionary<string, string>? ServiceNameMappings { get; init; }
+
+        public bool Equals(ImmutableDynamicSettings? other)
+        {
+            if (ReferenceEquals(null, other))
+            {
+                return false;
+            }
+
+            if (ReferenceEquals(this, other))
+            {
+                return true;
+            }
+
+            return
+                RuntimeMetricsEnabled == other.RuntimeMetricsEnabled
+             && DataStreamsMonitoringEnabled == other.DataStreamsMonitoringEnabled
+             && CustomSamplingRules == other.CustomSamplingRules
+             && Nullable.Equals(GlobalSamplingRate, other.GlobalSamplingRate)
+             && SpanSamplingRules == other.SpanSamplingRules
+             && LogsInjectionEnabled == other.LogsInjectionEnabled
+             && AreEqual(HeaderTags, other.HeaderTags)
+             && AreEqual(ServiceNameMappings, other.ServiceNameMappings);
```

Comment:
These will never be equal, right, because you create a new Dictionary every time in `OnConfigurationChanged`? 🤔 Isn't that problematic for checking `dynamicSettings.Equals(oldSettings.DynamicSettings)`? Do you have a unit test for this?

Reply:
Well that's why `AreEqual` compares the keys and values of the dictionaries